src/Elastic.Apm/CurrentExecutionSegmentsContainer.cs

```diff
@@ -25,7 +31,13 @@ public Span CurrentSpan
 		public Transaction CurrentTransaction
 		{
 			get => _currentTransaction.Value;
-			set => _currentTransaction.Value = value;
+			set
+			{
+				_currentTransaction.Value = value;
+				if (!_transactionObservers.IsValueCreated) return;
+
+				foreach (var observer in TransactionObservers) observer?.ActiveTransactionChanged(value);
```

Comment:
This will throw if `_transactionObservers` is changed either on another thread or on the same thread by one of the `ITransactionObserver`-s. One way to fix it would be to make `_transactionObservers` immutable by "copy-on-write" approach:
- Make `_transactionObservers` `volatile`
- Before read-only access: Making a local copy of `_transactionObservers` reference before iterating over it - we can assume that in this case read-only access is much more frequent operation and reading `volatile` when there're not frequent writes should be cheap
- Before changing:  Under lock making a copy of the list and replacing `_transactionObservers` reference. Lock is necessary in addition to `volatile` to prevent loosing updates from contending threads
Since `_transactionObservers` will not be `readonly` anymore it will make it impossible to use `Lazy` but it's not that much of a loss because initial value is just an empty `List` and it's not that expensive to create eagerly.

Reply:
+1 for catching this another option is to switch to `ConcurrentBag<T>` whos `GetEnumerator()` is thread safe and represents a point in time copy once called: https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1.getenumerator?redirectedfrom=MSDN&view=netframework-4.8#System_Collections_Concurrent_ConcurrentBag_1_GetEnumerator

`ConcurrentBag` is pretty slow for large collections but the expectations is that we'd only have a handful of observers at maximum

Reply:
By looking at [`ConcurrentBag.GetEnumerator()` implementation](https://referencesource.microsoft.com/#system/sys/system/collections/concurrent/ConcurrentBag.cs,585) it doesn't look very cheap for read-only access (takes lock, allocates and copies the elements, etc.).

Reply:
In any case it's worth deciding on https://github.com/elastic/apm-agent-dotnet/pull/636#pullrequestreview-326615287 first since it might render `ITransactionObserver` unnecessary.

Reply:
BTW both approaches ("copy-on-write" `List` and `ConcurrentBag`) still have some nasty corner cases - for example with some unfortunate thread scheduling some threads might still hold an older cached copy causing inconsistency between threads (although for at most one notification cycle). In this case it's most likely won't be much of an issue because all the `ITransactionObserver`-s are added before transactions start.

Reply:
Aye, was not arguing that either solved all threading issues,  `ConcurrentBag<T>` is a cheap solution in terms of development time now while we validate all the logging integrations. 

The preferred way is not to rely on these observers at all and plug into the render pipeline if at all possible (nlog/serilog) won't require this infrastructure. 

(quick comment need to board a flight now)

Reply:
We don’t really need a list here - there is no current need for it at the moment. So probably it wasn't a good idea from me anyway. 

We could just make `internal List<ITransactionObserver> TransactionObservers` to `internal ITransactionObserver`. And from that point I’d not really stress about synchronization - if the `ITransactionObserver` is set during a call to this method, it’s fine if it’s called, but it’s also fine if it doesn’t - it’ll be the next time.